Re ordering instanced pipelines

EstebanFS opened the pull request:
## What does this PR accomplish?
<!--
Choose all that apply.
Also, mention the linked issue here.
This will magically close the issue once the PR is merged.
-->
Bug Fix | **Feature** | Documentation

closes #6617 

## Changes proposed by this PR:
- Add functionality to allow instanced pipelines to be re-ordered within their group from the UI/fly

## Release Note

* Instanced Pipelines are allowed to be re-ordered with in their group through the UI (using the drag and drop functionality) or using the fly command: 
`fly -t dev oip -g groupName -p key1:var1 -p key2:var2`

## Contributor Checklist
- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [x] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

EstebanFS committed:
B: add migration for secondary ordering

concourse/concourse#6617

we populate this column with an ordering based on how instanced
pipelines are currently ordered within an instance group - i.e. by ID

Signed-off-by: Esteban Foronda <eforonda@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: allow ordering within instance group in DB

when we save an instanced pipeline, we set the secondary_ordering to the
previous secondary_ordering + 1. we use secondary_ordering for, you
guessed it, ordering pipelines when we return them via the API

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
B: Add end-point for ordering pipelines in group

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
S: Use teamHandleFactory for ordering API

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
B: add Effect for re-ordering instanced pipelines

...in the UI

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
S: Drag/DropState keeps track of Card

...instead of cardIdentifier and teamName.

previously, we were using cardIdentifier to keep track of what card we
were dragging/dropping over. this makes it harder to access information
about the cards - for instance, when dragging an InstancedPipelineCard,
we don't have an easy way of figuring out what instance group the card
belongs to - we'd have to either add the pipeline name on DragState
redundantly, or lookup the card by its identifier.

storing the full card that we are dragging makes this easier, since no
lookup is required. it also lets us remove the teamName on the
DragState, since that information can be inferred from the card

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
B: support drag and drop for instanced pipelines

we needed to change how we displayed Group headers - previously, we were
just using teamName as the header text. for instance group headers, we
previously just made the `teamName` the header we wanted
(i.e. teamName ++ " / " ++ groupName). this was a pretty bad hack,
because we rely on comparing the teamName for some drag and drop logic.

to fix this, we created a new type "Section" which allows for specifying
a header that differs from the team name. this way, instance groups can
have a header like before (teamName / groupName) while still using the
correct team name for comparisons.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

EstebanFS committed:
B: Add fly order-instance-pipelines command

Signed-off-by: Taylor Silva <tsilva@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: specify group name for order-instanced-pipelines

rather than needing to specify the pipeline name multiple times:

fly oip -p group/k1:v1 -p group/k2:v2

...you now specify the group name once and the list of instance vars
separately:

fly oip -g group -p k1:v1 -p k2:v2

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
B: quote / in var reference

we quoted / when it appeared in an instance var's value, but not when it
appeared in the key. while we still parse var references properly if
they contain a / without quoting, when we generate the string, we should
quote it for consistency

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
bump migration

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
remove transaction from migration

concourse/concourse#6727 wraps all migrations in a transaction, and we
can't start a transaction within another

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
fix yarn analyse

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

aoldershaw committed:
add error check in migration test

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

EstebanFS committed:
Fix pre and post migration version

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

EstebanFS committed:
Regenerate fakeTeam

Signed-off-by: Esteban Foronda <eforonda@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

EstebanFS committed:
Unfocus fly test, Fix execute_test

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

chenbh commented on fly/commands/ordering_instanced_pipeline.go:
```diff
@@ -0,0 +1,58 @@
+
+type OrderInstancedPipelinesCommand struct {
+	Pipeline     string                         `short:"g" long:"group" required:"true" description:"Name of the instance group"`
+	InstanceVars []flaghelpers.InstanceVarsFlag `short:"p" long:"pipeline" required:"true" description:"Instance vars identifying pipeline to order"`
```
This can also be applied to `OrderPipelinesCommand`

```suggestion
	InstanceVars []flaghelpers.InstanceVarsFlag `short:"p" long:"pipeline" required:"true" description:"Instance vars identifying pipeline (can be specified multiple times to provide relative ordering)"`
```

chenbh commented on fly/commands/internal/flaghelpers/instance_vars_flag.go:
```diff
@@ -0,0 +1,108 @@
+	InstanceVars atc.InstanceVars
+}
+
+func (flag *InstanceVarsFlag) UnmarshalFlag(value string) error {
+	var err error
+	flag.InstanceVars, err = unmarshalInstanceVars(value)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
Is it weird that we use `key=value` when setting a pipeline and `key:value` when referencing the pipeline?

aoldershaw replied:
The `:`/`=` discrepancy was mentioned in the initial RFC: https://github.com/concourse/rfcs/blob/master/034-instanced-pipelines/proposal.md

> (Side note: `:` vs. `=` is a little weird but it's consistent with `fly check-resource` - we use `=` for assignment and `:` for partial filtering.)

aoldershaw replied:
Moved most of the pipeline flag tests out to test just the instance vars flag

chenbh commented on web/elm/src/Dashboard/Dashboard.elm:
```diff
@@ -675,24 +683,42 @@ update session msg =
     case msg of
-        DragStart teamName cardId ->
-            ( { model | dragState = Models.Dragging teamName cardId }, effects )
+        DragStart card ->
+            ( { model | dragState = Models.Dragging card }, effects )
```
Don't think it should be fixed in this PR, but I noticed a bug where the card doesn't render if I drag the boxes inside the card (i.e the jobs for pipelines or pipelines for instance groups).

https://user-images.githubusercontent.com/17577073/115045536-9796ce00-9ea4-11eb-8b4a-d5eefde737c5.mov

chenbh commented:
Tested both via UI and fly, everything looks good and just some minor comments

aoldershaw commented on web/elm/src/Api/Endpoints.elm:
```diff
@@ -297,3 +309,12 @@ teamEndpoint endpoint =
     )
+
+
+instanceGroupEndpoint : InstanceGroupEndpoint -> RouteBuilder
+instanceGroupEndpoint endpoint =
+    ( case endpoint of
+        OrderInstanceGroupPipelines ->
+            [ "ordering" ]
+    , []
+    )
```
`pipelineEndpoint` identifies individual pipelines via `teamName`, `pipelineName`, *and* `pipelineInstanceVars` - `instanceGroupEndpoint` is only used on instance groups, so has no `pipelineInstanceVars`

aoldershaw commented on atc/db/team.go:
```diff
@@ -431,22 +430,25 @@ func savePipeline(
 			values["ordering"] = ordering.Int64
+			values["secondary_ordering"] = secondaryOrdering.Int64 + 1
 		} else {
 			values["ordering"] = sq.Expr("currval('pipelines_id_seq')")
+			values["secondary_ordering"] = 1
 		}
```
Not really, we just need a consistent ordering here and it felt more natural that way, Also Ordering is not 0-base at the beginning it only change his ordering to 0-base after re-ordering some pipelines.

aoldershaw replied:
To clarify the second part - we just take `currval('pipelines_id_seq')`, which will leave you with some value of `ordering` that is larger than any other pipeline currently.

`secondary_ordering` on the other hand is easy to keep as 1-based, and felt more natural to do so (since the `row_number` function we use in the migration is also 1-based, and it'd be a bit awkward to subtract 1 when all you really care about is the relative ordering).

That said, there isn't really a good reason for it - we're happy to reconsider this, I don't think either @EstebanFS or I feel particularly strongly about it one way or another

chenbh replied:
I don't particularly care either, I guess it's just odd that what is (essentially) a list starting at `1`

aoldershaw committed:
Solve Daniel's comments

Log team name on atc/api
Test instance vars flag
Make missing pipeline error more friendly on fly
Update help text
Fixing migration order

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>
Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented on go-concourse/concourse/pipelines.go:
```diff
@@ -57,6 +57,28 @@ func (team *team) OrderingPipelines(pipelineNames []string) error {
 	}, &internal.Response{})
 }
 
+func (team *team) OrderingPipelinesWithinGroup(groupName string, instanceVars []atc.InstanceVars) error {
```
We switched go-concourse to use the now preferred `HttpAgent` rather than the `Connection`, and special case `http.StatusBadRequest` so that we can display a more friendly error message

clarafu commented on fly/commands/ordering_instanced_pipeline.go:
```diff
@@ -0,0 +1,58 @@
+
+type OrderInstancedPipelinesCommand struct {
+	Pipeline     string                         `short:"g" long:"group" required:"true" description:"Name of the instance group"`
+	InstanceVars []flaghelpers.InstanceVarsFlag `short:"p" long:"pipeline" required:"true" description:"Instance vars identifying pipeline (can be specified multiple times to provide relative ordering)"`
```
Was there a reason we wanted to have the flag name be `"pipeline"` instead of `"instance_var"` or `"i"` for short? I'm asking because for the `set-pipeline` command we use `"instance-var"` as the flag name where you pass in the instance vars that are to be set https://github.com/concourse/concourse/blob/34474b0f0cd5b6a46639b517e4b38d3d4250f6e7/fly/commands/set_pipeline.go#L29
With how the `set-pipeline` command is used, I thought that a `pipeline` can have multiple `instances` but for this `order-instance-pipelines` command it looks like a `group` (equivalent to `pipeline`) can have multiple `pipelines` (equivalent to `instances`). Just wondering what your thoughts are!

EstebanFS replied:
We tried to leave the call to this command as similar as possible to the [order-pipeline](https://concourse-ci.org/managing-pipelines.html#fly-order-pipelines) command (so I could be friendly for the user), that's why we left the `pipeline` name flag to identify each instance pipeline and we added another flag called `group` to identify pipeline.

We checked also the syntax of the [set-pipeline](https://concourse-ci.org/setting-pipelines.html#fly-set-pipeline) command, but we though it could be more confusing for the user that way, because of the way [order-pipeline](https://concourse-ci.org/managing-pipelines.html#fly-order-pipelines) command was created.

Also we didn't want it to match [set-pipeline](https://concourse-ci.org/setting-pipelines.html#fly-set-pipeline) because `-i` it's used in a different way here, `-i` in [set-pipeline](https://concourse-ci.org/setting-pipelines.html#fly-set-pipeline) always sets a single instance var using `KEY=VALUE`, while `-p` in `order-instanced-pipelines` uses `KEY1:VALUE1,KEY2:VALUE2,...`
which could introduce its own confusion, because `-p` is usually `GROUP/KEY1:VALUE1,KEY2:VALUE2,...,` while here we omit the `GROUP/` because it's redundant, but still, it's closer to `-p` in other commands than `-i` in set-pipeline, we think.

Also, you're conceptually ordering pipelines (`-p`), not ordering instance vars (`-i`)

clarafu replied:
I see, I think I understand! I always saw instance pipelines as a subset of a pipeline, but I guess I should more be thinking about it as an instance pipeline is equal to just a pipeline and pipelines can be grouped. Thanks for the explanation!

I think a small suggestion I have is to rename the `Pipeline` field to `Group` so that it is clear that the field is referencing a group rather than a pipeline?

EstebanFS replied:
That's right, I just made the change there

clarafu commented on atc/db/team.go:
```diff
@@ -732,6 +734,54 @@ func (t *team) OrderPipelines(names []string) error {
 	return nil
 }
 
+func (t *team) OrderPipelinesWithinGroup(groupName string, instanceVars []atc.InstanceVars) error {
```
I guess this is more of a general question about ordering in general, but I was wondering what happens if you only order a few instances within a pipeline? For example, if I have 3 instances within a pipeline
```
instance1 -> secondary_ordering=1
instance2 -> secondary_ordering=2
instance3 -> secondary_ordering=3
```

And I run `fly -t <target> order-instance-pipelines -g <group> -p instance3 -p instance2`, it will set the `secondary_ordering` for `instance2` and `instance3`.
```
instance1 -> secondary_ordering=1
instance2 -> secondary_ordering=2
instance3 -> secondary_ordering=1
```

It will result in both `instance1` and `instance3` having `secondary_ordering` of 1, and in the UI I would see the ordering to be instance1 -> instance3 -> instance2. Do we want to just ignore the ordering of instances that are not passed into the fly command ordering? I know this is also a problem with the regular `fly order-pipelines` so it might not be necessary to address this here.

chenbh replied:
My opinion of it is that the `order-[instance-]pipelines` command allows you to specify the relative ordering between pipelines you care about. If you don't include it in the list, then you probably don't care about where it actually appears.

But then this is also my opinion after reading the docs, this could be a case of "it's not a bug, it's a feature!"

EstebanFS replied:
Exactly

clarafu replied:
Oh ya I see that comment in the doc haha Does seem weird to keep it like that but honestly its not a big deal 😄

clarafu reviewed (commented):
Tested it out manually, with fly commands and reordering in the UI and it all looks great!! :) All the code looks 👍  too :D

chenbh reviewed (approved):
lgtm

EstebanFS committed:
Rename Pipeline to Group

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>

clarafu reviewed (approved):
NOICE!
